File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/J._Случай_верблюда/solution.py
from __future__ import annotations
import sys


class Trie:
    def __init__(self) -> None:
        self.tree: dict[int, Trie] = dict()
        self.words: dict[str, int] = dict()

    def insert(self, word: str) -> None:
        root: Trie = self
        acronym: str = self.get_acronym(word)
        if not acronym:
            if 91 not in root.tree:
                root.tree[91] = Trie()
                root.tree[91].words[word] = root.tree[91].words.get(word, 0) + 1
                return
        for letter in acronym:
            if ord(letter) not in root.tree:
                root.tree[ord(letter)] = Trie()
            root = root.tree[ord(letter)]
        root.words[word] = root.words.get(word, 0) + 1

    def get_acronym(self, word: str) -> str:
        return ''.join(letter for letter in word if letter.isupper())

    def fill_set(self, root: Trie, commands: list[tuple[str, int]]) -> None:
        for word, freq in root.words.items():
            commands.append((word, freq))
        for idx in range(65, 92):
            if idx not in root.tree: continue
            self.fill_set(root.tree[idx], commands)

    def print_command(self, command: str) -> None:
        root: Trie = self
        for letter in command:
            if ord(letter) not in root.tree: return
            root = root.tree[ord(letter)]
        commands: list[tuple[str, int]] = []
        self.fill_set(root, commands)
        commands.sort()
        for word, freq in commands:
            sys.stdout.write('\n'.join(word for _ in range(freq)))
            sys.stdout.write('\n')


def solution() -> None:
    """
    Time Complexity O(N + MNlog(N))
    Memory Complexity O(L)
    """
    n: int = int(sys.stdin.readline().rstrip())
    trie: Trie = Trie()
    for _ in range(n):
        word: str = sys.stdin.readline().rstrip()
        trie.insert(word)
    m: int = int(sys.stdin.readline().rstrip())
    for _ in range(m):
        command: str = sys.stdin.readline().rstrip()
        trie.print_command(command)


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/J._Случай_верблюда/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <map>

class Trie {
private:
    std::unordered_map<int, Trie*> tree;
    std::map<std::string, int> words;
public:
    Trie() {}

    void insert(std::string& word) {
        Trie* root = this;
        std::string acronym = getAcronim(word);
        if (acronym.empty()) {
            root->tree[91] = new Trie();
            root->tree[91]->words[word] += 1;
            return;
        }
        for (char& letter : acronym) {
            if (root->tree.count((int)letter) == 0) {
                root->tree[(int)letter] = new Trie();
            }
            root = root->tree[(int)letter];
        }
        ++root->words[word];
    }

    std::string getAcronim(std::string& word) {
        std::string output = "";
        for (char& letter : word) {
            if (std::isupper(letter)) output.push_back(letter);
        }
        return output;
    }

    void printQuery(std::string& query) {
        Trie* root = this;
        for (char& letter : query) {
            if (root->tree.count((int)letter) == 0) {
                return;
            }
            root = root->tree[(int)letter];
        }
        std::map<std::string, int> commands;
        fillSet(root,commands);
        for (const auto& p : commands) {
            for (int i = 0; i < p.second; ++i) std::cout << p.first << "\n";
        }
    }

    void fillSet(Trie* root, std::map<std::string, int>& commands) {
        for (const auto& p : root->words) {
            commands[p.first] = p.second;
        }
        for (int idx = 65; idx <= 91; ++idx) {
            if (root->tree.count(idx)) fillSet(root->tree[idx], commands);
        }
    }
};

void solution() {
    /*
    Time Complexity O(N + MNlog(N))
    Memory Complexity O(L)
    */
    int n;
    std::scanf("%d", &n);
    Trie trie;
    for (int i = 0; i < n; ++i) {
        std::string command;
        std::cin >> command;
        trie.insert(command);
    }
    int m;
    std::scanf("%d", &m);
    for (int i = 0; i < m; ++i) {
        std::string query;
        std::cin >> query;
        trie.printQuery(query);
    }
    std::printf("\n");
}

int main() {
    solution();
    return 0;
}
